azure/services/securitygroups/securitygroups_test.go

```diff
@@ -22,213 +22,131 @@ import (
-									SourcePortRange:          to.StringPtr("*"),
-									DestinationPortRange:     to.StringPtr("*"),
-									SourceAddressPrefix:      to.StringPtr("*"),
-									DestinationAddressPrefix: to.StringPtr("*"),
-									Protocol:                 "*",
-									Direction:                "Inbound",
-									Access:                   "Allow",
-									Priority:                 to.Int32Ptr(400),
-								},
-								Name: to.StringPtr("first-rule"),
-							},
-							{
-								SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
-									Description:              to.StringPtr("another test rule"),
-									SourcePortRange:          to.StringPtr("*"),
-									DestinationPortRange:     to.StringPtr("*"),
-									SourceAddressPrefix:      to.StringPtr("*"),
-									DestinationAddressPrefix: to.StringPtr("*"),
-									Protocol:                 "*",
-									Direction:                "Inbound",
-									Access:                   "Allow",
-									Priority:                 to.Int32Ptr(450),
-								},
-								Name: to.StringPtr("second-rule"),
-							},
-						},
-					},
-					Etag:     nil,
-					Location: to.StringPtr("test-location"),
-				}))
-				m.Get(gomockinternal.AContext(), "my-rg", "nsg-two").Return(network.SecurityGroup{}, autorest.NewErrorWithResponse("", "", &http.Response{StatusCode: 404}, "Not found"))
-				m.CreateOrUpdate(gomockinternal.AContext(), "my-rg", "nsg-two", gomockinternal.DiffEq(network.SecurityGroup{
-					SecurityGroupPropertiesFormat: &network.SecurityGroupPropertiesFormat{
-						SecurityRules: &[]network.SecurityRule{},
-					},
-					Etag:     nil,
-					Location: to.StringPtr("test-location"),
-				}))
+				s.IsVnetManaged().Return(true, nil)
+				s.NSGSpecs().Return(fakeNSGSpecs)
+				s.GetLongRunningOperationState("test-nsg", serviceName).Return(nil)
+				m.CreateOrUpdateAsync(gomockinternal.AContext(), &fakeNSG).Return(nil, errFake)
+				s.GetLongRunningOperationState("test-nsg-2", serviceName).Return(nil)
+				m.CreateOrUpdateAsync(gomockinternal.AContext(), &fakeNSG2).Return(nil, nil)
+				s.UpdatePutStatus(infrav1.SecurityGroupsReadyCondition, serviceName, gomockinternal.ErrStrEq("failed to create resource test-group/test-nsg (service: securitygroups): this is an error"))
 			},
-		}, {
-			name: "security group exists",
+		},
+		{
+			name:          "second security groups create not done",
+			expectedError: "failed to create resource test-group/test-nsg (service: securitygroups): this is an error",
+			expect: func(s *mock_securitygroups.MockNSGScopeMockRecorder, m *mock_securitygroups.MockclientMockRecorder) {
+				s.V(gomock.AssignableToTypeOf(2)).AnyTimes().Return(klogr.New())
+				s.IsVnetManaged().Return(true, nil)
+				s.NSGSpecs().Return(fakeNSGSpecs)
+				s.GetLongRunningOperationState("test-nsg", serviceName).Return(nil)
+				m.CreateOrUpdateAsync(gomockinternal.AContext(), &fakeNSG).Return(nil, errFake)
```

Comment:
@CecileRobertMichon Looking at the name of the test `second security groups create fails`, I'm wondering if it matches the implementation. The implementation has an error for the creation of both the security groups and the result status is from the first failure. From the test name I was thinking the implementation might look like

first security group `CreateOrUpdateAsync` - passes, no errors, no futures as it's immediately created. But currently it's different

second security `CreateOrUpdateAsync` - has errors but it has future as it's create is not yet done. This is how it's currently done, except that `fakeFuture` has reference to `DELETE` HTTP Method - https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1684/files#diff-5bfc905db237a71559b05fbd32c6ec3f216c5b5060b8e1824abc27b97fddaab2R76

Also, generic question, looks like `CreateOrUpdateAsync` is [expected to return a non-nil error even when there's a future](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/async/async.go#L77-L79). Why is that the case? I see that `WaitForCompletionRef` used in the async implementations can only fail due to context cancellation (in our case due to timeout) or other timing reasons (exceeds polling duration etc). If that's the case, why not return nil error along with future? I'm asking as it's weird to see an error being returned in the test called `errFoo` and the only reason is because `async.CreateResource` expects `CreateOrUpdateAsync` to return error along with future which is what we do here https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1684/files#diff-53d4a612701b97cfb63f5e584b0e1392ee162a7bee333063721cccb1604b4c8dR109 . The first thought the came to mind was - why not return future with nil error when it's sure that there's no error other than "operation not done" or "operation pending" error

Reply:
updated the test names to make what they're testing more clear. Also added a fake PUT future to use for reconcile tests. 

> why not return future with nil error when it's sure that there's no error other than "operation not done" or "operation pending" error
addressed in #1837